src/libstore/gc.cc

```diff
@@ -336,19 +337,115 @@ Roots LocalStore::findRoots()
     return roots;
 }
 
+static void tryRuntimeRootLink(Path path, bool ifAbsolute, StringSet & paths) {
+    try {
+        auto target = readLink(path);
+        if (!ifAbsolute || target.find("/") == 0) {
+            paths.insert(target);
+        }
+    } catch (...) { }
+}
+
+static void findRuntimeRootsInProc(StringSet & paths) {
+    if (getFileType("/proc") != DT_DIR) return;
+
+    auto procreg = Regex("[0-9]+", false);
+    auto mapreg = Regex("[ ]*[^ ]+[ ]+[^ ]+[ ]+[^ ]+[ ]+[^ ]+[ ]+[^ ]+[ ]+(/[^ ]+)[ ]*", true);
+    auto pathreg = Regex("/[0-9a-z]+[0-9a-zA-Z+._?=-]*", true);
+
+    for (auto & i : readDirectory("/proc")) {
+        if (!procreg.matches(i.name)) continue;
+
+        auto process = "/proc/"+i.name;
+        tryRuntimeRootLink(process+"/exe", false, paths);
+        tryRuntimeRootLink(process+"/cwd", false, paths);
+
+        try {
+            for (auto & i : readDirectory(process+"/fd")) {
+                tryRuntimeRootLink(process+"/fd/"+i.name, true, paths);
+            }
+        } catch (...) { }
+
+        try {
+            auto result = readFile(process+"/maps", true);
+            auto lines = tokenizeString<Strings>(result, "\n");
+
+            foreach (Strings::iterator, i, lines) {
+                Regex::Subs subs;
+                if (mapreg.matches(*i, subs)) {
+                    paths.insert(subs[0]);
+                }
+            }
+        } catch (...) { }
+
+        try {
+            auto environ = readFile(process+"/environ", true);
+            size_t pos = 0;
+            while ((pos = environ.find(settings.nixStore, pos)) != string::npos) {
+                pos += settings.nixStore.size();
```

Comment:
This can cause an attempt to read beyond the end of the string.

Also, it appears that reading roots from `environ` doesn't work properly. When I do a diff between the roots found by the old and new runtime root finder, the latter lacks numerous paths that only occur in `environ`.